service/test/BatchServerTest.cpp

```diff
@@ -359,6 +461,263 @@ TEST_F(BatchServerTest, run_batch_write_empty)
+ */
+ACTION(batch_status_write_EINTR)
+{
+    throw geopm::Exception(
+        "BatchStatusImp: System call failed: " "write(2)",
+        EINTR,
+        __FILE__,
+        __LINE__
+    );
+}
+
+/**
+ * @test Testing BatchStatusImp::send_message() to throw an exception inside of BatchServerImp::write_message()
+ *       This is needed to activate the catch blocks in BatchServerImp::write_message() and BatchServerImp::run_batch()
+ */
+TEST_F(BatchServerTest, write_message_exception)
+{
+    InSequence sequence;
+
+    int idx = 0;
+    std::vector<double> result = {240.042, 250.052};
+
+    for (const auto &request : m_signal_config) {
+        EXPECT_CALL(*m_pio_ptr, push_signal(request.name, request.domain,
+                                            request.domain_idx))
+            .WillOnce(Return(idx))
+            .RetiresOnSaturation();
+        ++idx;
+    }
+
+    for (const auto &request : m_control_config) {
+        EXPECT_CALL(*m_pio_ptr, push_control(request.name, request.domain,
+                                             request.domain_idx))
+            .WillOnce(Return(idx))
+            .RetiresOnSaturation();
+        ++idx;
+    }
+
+    EXPECT_CALL(*m_batch_status, receive_message())
+        .WillOnce(Return(BatchStatus::M_MESSAGE_QUIT))
+        .RetiresOnSaturation();
+
+    // This is where it throws the message.
+    EXPECT_CALL(
+        *m_batch_status.get(),
+        send_message(BatchStatus::M_MESSAGE_QUIT)
+    )
+    .Times(1)
+    .WillRepeatedly(batch_status_write_EINTR());
+
+    EXPECT_CALL(*m_batch_status,
+                send_message(BatchStatus::M_MESSAGE_QUIT))
+        .Times(1)
+        .RetiresOnSaturation();
+
+    GEOPM_EXPECT_THROW_MESSAGE(
+        m_batch_server->run_batch(),
+        EINTR,
+        "BatchStatusImp: System call failed: " "write(2)"
+    );
+}
+
+/**
+ * @throws geopm Exception to simulate failure of ioctl(2) system call in SSTIoctlImp::mbox() in SSTIOImp::read_batch()
+ */
+ACTION(read_batch_mbox_failed_EINVAL)
+{
+    throw geopm::Exception(
+        "SSTIOImp::read_batch() mbox read failed",
+        EINVAL,
+        __FILE__,
+        __LINE__
+    );
+}
+
+/**
+ * @test Testing SSTIOImp::read_batch() to throw an exception inside of BatchServerImp::read_and_update()
+ *       This is needed to activate if (m_is_client_waiting) branch inside the catch block in BatchServerImp::run_batch()
+ */
+TEST_F(BatchServerTest, read_batch_exception)
+{
+    InSequence sequence;
+
+    int idx = 0;
+    std::vector<double> result = {240.042, 250.052};
+
+    for (const auto &request : m_signal_config) {
+        EXPECT_CALL(*m_pio_ptr, push_signal(request.name, request.domain,
+                                            request.domain_idx))
+            .WillOnce(Return(idx))
+            .RetiresOnSaturation();
+        ++idx;
+    }
+
+    for (const auto &request : m_control_config) {
```

Comment:
```suggestion
    idx = 0;
    for (const auto &request : m_control_config) {
```

When pushing signals and controls, make sure the index is reset in between.  It doesn't matter for this test because you are not verifying the indices, but it will lead to confusing behavior if this code is used in other tests that do utilizes the indices.